conformance/tests/directives_deprecated.py

```diff
@@ -0,0 +1,108 @@
+"""
+Tests the warnings.deprecated function.
+"""
+
+# pyright: reportDeprecated=true
+
+# Specification: https://typing.readthedocs.io/en/latest/spec/directives.html#deprecated
+# See also https://peps.python.org/pep-0702/
+
+# > Type checkers should produce a diagnostic whenever they encounter a usage of an object
+# > marked as deprecated. [...] For deprecated classes and functions, this includes:
+
+# > * `from` imports
+
+from _directives_deprecated_library import Ham  # E: Use of deprecated class Ham
+import _directives_deprecated_library as library
+
+from typing_extensions import deprecated
+
+
+# > * References through module, class, or instance attributes
+
+library.norwegian_blue(1)  # E: Use of deprecated function norwegian_blue
+map(library.norwegian_blue, [1, 2, 3])  # E: Use of deprecated function norwegian_blue
+
+
+# > For deprecated overloads, this includes all calls that resolve to the deprecated overload.
+
+library.foo(1)  # E: Use of deprecated overload for foo
+library.foo("x")  # no error
+
+
+ham = Ham()  # no error (already reported above)
+
+
+# > Any syntax that indirectly triggers a call to the function.
```

Comment:
Another useful test is to verify that a `__call__` marked as deprecated (e.g. within a callback protocol) triggers a deprecation check if the object is called.